src/backend/distributed/commands/dependencies.c

```diff
@@ -0,0 +1,216 @@
+
+
+/*
+ * EnsureDependenciesExists finds all the dependencies that we support and makes sure
+ * these are available on all workers. If not available they will be created on the
+ * workers via a separate session that will be committed directly so that the objects are
+ * visible to potentially multiple sessions creating the shards.
+ *
+ * Note; only the actual objects are created via a separate session, the local records to
+ * pg_dist_object are created in this session. As a side effect the objects could be
+ * created on the workers without a catalog entry on the coordinator. Updates to the
+ * objects on the coordinator are not propagated to the workers until the record is
+ * visible on the coordinator.
+ *
+ * This is solved by creating the dependencies in an idempotent manner, either via
+ * postgres native CREATE IF NOT EXISTS, or citus helper functions.
+ */
+void
+EnsureDependenciesExistsOnAllNodes(const ObjectAddress *target)
+{
+	const uint64 connectionFlag = FORCE_NEW_CONNECTION;
+
+	/* local variables to work with dependencies */
+	List *dependencies = NIL;
+	ListCell *dependencyCell = NULL;
+
+	/* local variables to collect ddl commands */
+	List *ddlCommands = NULL;
+
+	/* local variables to work with worker nodes */
+	List *workerNodeList = NULL;
+	ListCell *workerNodeCell = NULL;
+	List *connections = NULL;
+	ListCell *connectionCell = NULL;
+
+	/*
+	 * collect all dependencies in creation order and get their ddl commands
+	 */
+	dependencies = GetDependenciesForObject(target);
+	foreach(dependencyCell, dependencies)
+	{
+		ObjectAddress *dependency = (ObjectAddress *) lfirst(dependencyCell);
+		ddlCommands = list_concat(ddlCommands,
+								  GetDependencyCreateDDLCommands(dependency));
+	}
+	if (list_length(ddlCommands) <= 0)
+	{
+		/* no ddl commands to be executed */
+		return;
+	}
+
+	/*
+	 * collect and connect to all applicable nodes
+	 */
+	workerNodeList = ActivePrimaryNodeList();
+	if (list_length(workerNodeList) <= 0)
+	{
+		/* no nodes to execute on */
+		return;
+	}
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		MultiConnection *connection = NULL;
+
+		char *nodeName = workerNode->workerName;
+		uint32 nodePort = workerNode->workerPort;
+
+		connection = StartNodeUserDatabaseConnection(connectionFlag, nodeName, nodePort,
+													 CitusExtensionOwnerName(), NULL);
+
+		connections = lappend(connections, connection);
+	}
+	FinishConnectionListEstablishment(connections);
+
+	/*
+	 * create dependency on all nodes
+	 */
+	foreach(connectionCell, connections)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+		ExecuteCriticalRemoteCommandList(connection, ddlCommands);
+	}
+
+	/*
+	 * mark all objects as distributed
+	 */
+	foreach(dependencyCell, dependencies)
+	{
+		ObjectAddress *dependency = (ObjectAddress *) lfirst(dependencyCell);
+		MarkObjectDistributed(dependency);
```

Comment:
If two sessions are creating the same type concurrently they will currently be allowed to insert the same record twice. If we add a primary key (which we should), then one of them will block and eventually fail on this line. Ideally, we would be doing a type of upsert here.

One small improvement would be to move these call to before where we do `ActivePrimaryNodeList()`, then at least the losing session is not going to try to redundantly create stuff on workers.

Reply:
This is might be a small regression from 8.3 as far as schemas are concerned. One fix could be to do an actual upsert via SPI.

We generally shun using SPI to accesss metadata for performance reasons, but for occasional writes it's fine.

Reply:
I think in general, if we create the target object locally first -- instead of doing it on the workers as of now, Postgres would block and error out accordingly. 

Another related item with the same topic: https://github.com/citusdata/citus/pull/2893/files#r316761726

I think we should also add this to TODO list that we'd want to handle later

Reply:
> I think in general, if we create the target object locally first -- instead of doing it on the workers as of now, Postgres would block and error out accordingly.

Not sure I follow. We're not in the DDL hook here, we're typically in create_distributed_table.